packages/connections/src/stores/connections-store.ts

```diff
@@ -0,0 +1,275 @@
+      void onConnected(
+        connectedConnectionInfo.current,
+        connectedDataService.current
+      );
+    }
+  }, [isConnected, onConnected]);
+
+  useEffect(() => {
+    async function loadConnections() {
+      try {
+        const connectionStorage = new ConnectionStorage();
+        const loadedConnections = await connectionStorage.loadAll();
+
+        dispatch({
+          type: 'set-connections',
+          connections: loadedConnections,
+        });
+      } catch (error) {
+        debug('error loading connections', error);
+      }
+    }
+
+    // Load connections after first render.
+    void loadConnections();
+
+    return () => {
+      // When unmounting, clean up any current connection attempts that have
+      // not resolved.
+      if (
+        connectingConnectionAttempt.current &&
+        !connectingConnectionAttempt.current.isClosed()
+      ) {
+        connectingConnectionAttempt.current.cancelConnectionAttempt();
+      }
+    };
+  }, []);
+
+  return [
+    state,
+    {
+      onCancelConnectionAttempt() {
+        connectionAttempt?.cancelConnectionAttempt();
+
+        dispatch({
+          type: 'cancel-connection-attempt',
+        });
+      },
+      async onConnect(connectionInfo: ConnectionInfo) {
+        if (connectionAttempt || isConnected) {
+          // Ensure we aren't currently connecting.
+          return;
+        }
+
+        const newConnectionAttempt = createConnectionAttempt();
+        connectingConnectionAttempt.current = newConnectionAttempt;
+
+        dispatch({
+          type: 'attempt-connect',
+          connectingStatusText: `Connecting to ${getConnectionTitle(
+            connectionInfo
+          )}`,
+          connectionAttempt: newConnectionAttempt,
+        });
+
+        debug('connecting with connectionInfo', connectionInfo);
+
+        try {
+          const newConnectionDataService = await newConnectionAttempt.connect(
+            connectionInfo.connectionOptions
+          );
+          connectingConnectionAttempt.current = undefined;
+
+          if (!newConnectionDataService || newConnectionAttempt.isClosed()) {
+            // The connection attempt was cancelled.
+            return;
+          }
+
+          // Successful connection
+          connectedConnectionInfo.current = connectionInfo;
+          connectedDataService.current = newConnectionDataService;
+
+          dispatch({
+            type: 'connection-attempt-succeeded',
+          });
+          debug(
+            'connection attempt succeeded with connection info',
+            connectionInfo
+          );
+        } catch (error) {
+          connectingConnectionAttempt.current = undefined;
+          debug('connect error', error);
+
+          dispatch({
+            type: 'connection-attempt-errored',
+            connectionErrorMessage: error.message,
+          });
+        }
+      },
+      updateActiveConnection(newConnectionId?: string | undefined) {
```

Comment:
Hyper-minor .. Could this be 2 different functions one to set the active connection and one to create a new one? Seems like the 2 are totally independent, dispatching different actions and called in different contexts. (we could still call one from another as a fallback if needed)

Reply:
Good call, updated.